web-ncer/app/services/SampleRequestService.scala

```diff
@@ -0,0 +1,298 @@
+ * @param dataSpaceMetaInfos Meta information
+ * @param elementGridWidth Grid width to use for charts
+ */
+case class ActionFormViewData(
+  dataViewId: BSONObjectID,
+  tableViewParts: Seq[TableViewData],
+  dataSetSetting: DataSetSetting,
+  dataSpaceMetaInfos: Traversable[DataSpaceMetaInfo],
+  elementGridWidth: Int
+)
+
+/**
+ * A REMS catalogue item
+ *
+ * @param id The id of the item
+ * @param name The readable name of the item
+ * @param formId The id associated with the catalogue item
+ */
+case class CatalogueItem(
+  id: Int,
+  name: String,
+  formId: Int
+)
+
+/**
+ * Service providing functionality to submit requests to REMS
+ */
+class SampleRequestService @Inject() (
+  dsaf: DataSetAccessorFactory,
+  sampleRequestSettingRepo: SampleRequestSettingRepo,
+  config: Configuration,
+  ws: WSClient,
+  val dataSpaceService: DataSpaceService
+) extends DataSetViewHelper {
+
+  private val remsUrl = config.getString("rems.url").getOrElse(
+    throw new AdaException("Configuration issue: 'rems.url' was not found in the configuration file.")
+  )
+  private val remsServiceUser = config.getString("rems.serviceUser").getOrElse(
+    throw new AdaException("Configuration issue: 'rems.serviceUser' was not found in the configuration file.")
+  )
+  private val remsUserPrefix = config.getString("rems.userPrefix").getOrElse(
+    throw new AdaException("Configuration issue: 'rems.userPrefix' was not found in the configuration file.")
+  )
+  private val remsMasterApiKey = config.getString("rems.masterApiKey").getOrElse(
+    throw new AdaException("Configuration issue: 'rems.masterApiKey' was not found in the configuration file.")
+  )
+
+  /**
+   * Create a valid CSV which can be attached to an REMS application
+   *
+   * @param dataSetId The data set id to use
+   * @param conditions The filter conditions. If empty no filter is applied
+   * @param fieldNames The field/column names to use. If empty all fields are selected
+   * @param selectedIds The selected row ids to use. If empty all are selected
+   * @return A string representing a valid CSV file.
+   */
+  def createCsv(
+    dataSetId: String,
+    conditions: Seq[FilterCondition],
+    fieldNames: Seq[String],
+    selectedIds: Seq[BSONObjectID]
+  ): Future[String] = {
+    val SEPARATOR = "\t"
+    val dsa = dsaf(dataSetId).getOrElse(throw new IllegalArgumentException(s"Dataset '$dataSetId' does not exist."))
+    val fieldCriteria = if (fieldNames.nonEmpty) Vector(FieldIdentity.name #-> fieldNames) else Nil
+    val selectCriteria = if (selectedIds.nonEmpty) Vector(JsObjectIdentity.name #-> selectedIds) else Nil
+    for {
+      fields <- dsa.fieldRepo.find(fieldCriteria)
+      valueCriteria <- FieldUtil.toDataSetCriteria(dsa.fieldRepo, conditions)
+      items <- dsa.dataSetRepo.find(valueCriteria ++ selectCriteria)
+    } yield {
+      val header = fields.map(_.name)
+      val csv = new StringBuilder("")
+      csv ++= header.mkString(SEPARATOR)
```

Comment:
this is quite a procedural way of creating csv... better to use map and mkString